move autocomplete list rendering to client side

johndoh opened the pull request:
Moving the rendering of the autocomplete address list used in compose to the client side allows skins to handle it and offer different layouts for different sized screens. Before I spend too much time on this I wanted to submit this draft to see if the idea was worth pursuing.

There are some hurdles. There is the `contact_search_name` config var which has to be respected. The server side needs to know about it so it can return the correct fields in the request but rendering those is a client side task because of screen size. Also `contact_search_name` does not support HTML and I think that's correct, it could be abused. May be having a configurable display for normal screens and a fixed one for small screens is enough.

Example of how the new small screen layout looks:
![compose](https://github.com/user-attachments/assets/19fa8ca2-86cc-4e42-87ce-bac89a426890)

The proposed `rcube_webmail.ksearch_results_display()` and `rcube_webmail.ksearch_results_highlight()` JS functions can be overridden by skins/plugins if required but the stock version maintain the current functionality plus a new layout for smaller screens.

I think ultimately all the text rendering code should be removed from `rcube_addressbook::compose_search_name()` and instead only the fields returned with the rendering moved into `app.js`.

This might help with #9329 and #9219.

johndoh committed:
basic support for autocomplete list rendering on client side

alecpl commented:
We should consider removing `contact_search_name` option. Maybe we can replace it with `autocomplete_extra_field` or something. This way the result would contain at most 3 items: "email", "display" (contact name), "extra" (e.g. department name). The extra text could be displayed not inline, but maybe aligned to the right, and only if there's room.

We can consider displaying "two line" records on a big screen too. Gmail does that.

johndoh commented:
I like the idea of scrapping `contact_search_name`. The `autocomplete_extra_field`, I think it would make sense if that could only be 0 or 1 fields and then skins have a fixed layout to support. If devs want more fields or more customisation then they could do it via the plugin API because its not only about the fields but also the layout. Does that sound acceptable?

Do we really need `autocomplete_extra_field` even, could `addressbook_name_listing` be used/improved so that if you wanted say `firstname lastname (department)` its shown the same in the autocomplete and the contact list? Then in autocomplete the email address would be on the second line and that's the only difference with the contacts list. It could use similar logic as that currently used for displaying `contact_search_name` so no existing configurability would be lost.

Off the top of my head I'm thinking a plugin hook in `rcube_addressbook::compose_search_name()` for the returned fields/values and then use the proposed (or similar) new JS functions to construct the entry.

alecpl commented:
Makes sense.

johndoh committed:
remove 'contact_search_name' config var, add 'rcube_addressbook::compose_autocomplete_fields()'

johndoh committed:
add contactlist_name_template config replacement for contact_search_name

johndoh committed:
elastic: style up autocomplete box

johndoh committed:
use rcube_addressbook::compose_autocomplete_fields() in acl plugin

johndoh committed:
cs fixes

johndoh commented:
I have replaced `contact_search_name` with `contactlist_name_template` which applies in the contact list and the autocomplete, anywhere `addressbook_name_listing` is used. I did not change `addressbook_name_listing` because of its use in the settings screen.

eef8919 I modified the ACL plugin to use the new autocomplete but I do not have a working LDAP setup so I cannot test this change, please could someone who uses this plugin confirm the autocomplete still works?

alecpl commented on plugins/acl/acl.php:
```diff
@@ -101,10 +101,10 @@ public function acl_autocomplete()
+                    $user = ['name' => $user, 'fields' => $fields];
                     $users[] = $user;
-                    $keys[] = $display ?: $user['name'];
+                    $keys[] = $fields['name'] ?: $user['name'];
```
There's no guarantee name is set.

alecpl commented on program/lib/Roundcube/rcube_addressbook.php:
```diff
@@ -763,55 +773,84 @@ public static function compose_search_name($contact, $email = null, $name = null
         static $template;
```
We don't need this static anymore. It was useful with `config->get()` as this can get slow, but you removed that call.

alecpl commented on program/lib/Roundcube/rcube_addressbook.php:
```diff
@@ -763,55 +773,84 @@ public static function compose_search_name($contact, $email = null, $name = null
+     */
+    public static function compose_autocomplete_fields($contact, $email = null, $name = null, $fields = null)
+    {
+        static $template;
```
This static is not needed anymore.

alecpl commented on program/lib/Roundcube/rcube_addressbook.php:
```diff
@@ -763,55 +773,84 @@ public static function compose_search_name($contact, $email = null, $name = null
+     *
+     * @return array Fields
+     */
+    public static function compose_autocomplete_fields($contact, $email = null, $name = null, $fields = null)
```
Should it be rather called `compose_search_fields()`? Also the description is wrong.

alecpl commented on program/lib/Roundcube/rcube_addressbook.php:
```diff
@@ -763,55 +773,84 @@ public static function compose_search_name($contact, $email = null, $name = null
-                    $value = strpos($key, ':') ? $contact[$key] : self::get_col_values($key, $contact, true);
-                    if (is_array($value) && isset($value[0])) {
-                        $value = $value[0];
+        $fields = $fields ?: $template;
```
You don't need `$template` variable at all. Just set `$fields` to the default here.

alecpl commented on program/lib/Roundcube/rcube_addressbook.php:
```diff
@@ -763,55 +773,84 @@ public static function compose_search_name($contact, $email = null, $name = null
+     * @param array  $contact Hash array with contact data as key-value pairs
+     * @param string $email   Optional email address
+     * @param string $name    Optional name (self::compose_list_name() result)
+     * @param array  $fields  Optional fields to return (defaults to ['name', 'email'])
```
Either allow null for $fields or not. It's incompatible with the function argument definition below. Same for `$name` and `$email`.

johndoh replied:
I dont understand

alecpl commented on program/lib/Roundcube/rcube_contacts.php:
```diff
@@ -347,7 +347,16 @@ public function search($fields, $value, $mode = 0, $select = true, $nocount = fa
                 $groups = [];
                 foreach ((array) $fields as $idx => $col) {
-                    $groups[] = $this->fulltext_sql_where($word, $mode, $col);
+                    // table column
```
This change does not seem to be related. Could be moved to a separate PR, with some explanation.

johndoh committed:
requested changes

johndoh committed:
check name is set

johndoh committed:
cs fix

johndoh commented:
I'm happy to work on it. Is there any particular changes you are looking for?

I already made the changes Alec requested and I think 3 failing tests are not related to this PR

pabzm commented:
@johndoh Maybe you could rebase onto the lastest of the "master" branch, it has this failing E2E-test fixed.

alecpl commented:
There's some bug here. If you have a contact that matches search string, but does not have any name, it will not be displayed (it's not returned by the search request).

johndoh committed:
restore 529e835

johndoh committed:
fic typo
